### PR TITLE
ShippingZones (GWY-5462)

### PR DESCRIPTION
**Pull request by dkotin**
Resolves: GWY-5462

### CODE REVIEW

**Comment by dkotin** [comment]
test this please

**Comment by dkotin** [comment]
retest this please

File: src/Entities/ShippingZone.php [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+ * For the full copyright and license information, please view the LICENSE.md
+ * file that was distributed with this source code.
+ */
+
+namespace Rebilly\Entities;
+
+use Rebilly\Rest\Entity;
+
+/**
+ * Class Attachment
```
Comment (XaosSintez) [inline src/Entities/ShippingZone.php]: Wrong PHPDoc

File: src/Entities/ShippingZone.php [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    /**
+     * @return $this
+     */
+    public function setRates($value)
+    {
+        return $this->setAttribute('rates', $value);
+    }
+
+    /**
+     * @return array
```
Comment (XaosSintez) [inline src/Entities/ShippingZone.php]: I doubt it

File: src/Entities/ShippingZone.php [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    {
+        return $this->getAttribute('rates');
+    }
+
+    /**
+     * @return $this
+     */
+    public function setRates($value)
+    {
+        return $this->setAttribute('rates', $value);
```
Comment (XaosSintez) [inline src/Entities/ShippingZone.php]: Shouldn't rates be a concrete object? I mean may be some validation and/or objects to compose an array?

File: src/Entities/ShippingZone.php [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    /**
+     * @return array
+     */
+    public function getIsDefault()
+    {
+        return $this->getAttribute('isDefault');
+    }
+
+    /**
+     * @return $this
```
Comment (XaosSintez) [inline src/Entities/ShippingZone.php]: @param missing

File: src/Services/ShippingZoneService.php [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    public function load($shippingZoneId, $params = [])
+    {
+        return $this->client()->get('shipping-zones/{shippingZoneId}', ['shippingZoneId' => $shippingZoneId] + (array) $params);
+    }
+
+    /**
+     * @param array|JsonSerializable|ShippingZone $data
+     * @param null $shippingZoneId
+     *
+     * @return ShippingZone The input data does not valid
```
Comment (XaosSintez) [inline src/Services/ShippingZoneService.php]: Wrong description

File: tests/Api/ApiTest.php [hunk 6]
```diff
@@ -945,6 +945,11 @@ public function provideServiceClasses()
                 Services\ValuesListTrackingService::class,
                 Entities\ValuesList::class,
             ],
+            [
+                'shippingZones',
+                Services\ShippingZoneService::class,
+                Entities\ShippingZone::class,
+            ],
```
Comment (XaosSintez) [inline tests/Api/ApiTest.php]: You are missing Entity test here

File: src/Entities/Shipping/Rate.php [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+            throw new DomainException(self::MSG_REQUIRED_CURRENCY);
+        }
+
+        $rate = new Rate($data);
+
+        return $rate;
+    }
+
+    /**
+     * @return mixed
```
Comment (XaosSintez) [inline src/Entities/Shipping/Rate.php]: Interesting, why `mixed` here and below? It should be float

File: src/Entities/Shipping/Rate.php [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+     */
+    public function getCurrency()
+    {
+        return $this->currency;
+    }
+
+    /**
+     * @param mixed $currency
```
Comment (XaosSintez) [inline src/Entities/Shipping/Rate.php]: And here should be string

File: src/Entities/Shipping/ShippingZone.php [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+    /**
+     * @return array
+     */
+    public function getRates()
+    {
+        return $this->getAttribute('rates');
+    }
+
+    /**
+     * @param []Rate $value
```
Comment (XaosSintez) [inline src/Entities/Shipping/ShippingZone.php]: Looks strange

File: src/Entities/Shipping/ShippingZone.php [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+    /**
+     * @return array
+     */
+    public function getCountries()
+    {
+        return $this->getAttribute('countries');
+    }
+
+    /**
+     * @param []string $value
```
Comment (XaosSintez) [inline src/Entities/Shipping/ShippingZone.php]: Looks strange

File: src/Entities/Shipping/ShippingZone.php [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+    /**
+     * @return bool
+     */
+    public function getIsDefault()
+    {
+        return $this->getAttribute('isDefault');
+    }
+
+    /**
+     * @param $value
```
Comment (XaosSintez) [inline src/Entities/Shipping/ShippingZone.php]: Add type here please

File: tests/Api/ApiTest.php [hunk 12]
```diff
@@ -743,6 +743,8 @@ public function provideEntityClasses()
             [Entities\Coupons\Coupon::class],
             [Entities\Coupons\Redemption::class],
             [Entities\ValuesList::class],
+            [Entities\Shipping\ShippingZone::class],
+            //[Entities\Shipping\Rate::class],
```
Comment (XaosSintez) [inline tests/Api/ApiTest.php]: Remove this please

File: tests/Api/ApiTest.php [hunk 13]
```diff
@@ -1278,6 +1285,29 @@ private function getFakeValue($attribute, $class)
                     $faker->word,
                     $faker->numberBetween(1, 100),
                 ];
+            case 'minOrderSubtotal':
+                return $faker->numberBetween(5, 7);
+            case 'maxOrderSubtotal':
+                return $faker->numberBetween(8, 10);
+
```
Comment (XaosSintez) [inline tests/Api/ApiTest.php]: Something strange here

File: tests/Api/ApiTest.php [hunk 14]
```diff
@@ -1278,6 +1285,29 @@ private function getFakeValue($attribute, $class)
...
+        case 'price':
+            return $faker->numberBetween(2, 3);
+
+        case 'countries':
+            return ['US'];
+        case 'isDefault':
+            return false;
+            case 'rates':
```
Comment (XaosSintez) [inline tests/Api/ApiTest.php]: And here

**Pull request merged**